graylog2-server/src/main/java/org/graylog2/security/UserSessionTerminationService.java

```diff
@@ -0,0 +1,219 @@
+    private final DefaultSecurityManager securityManager;
+    private final ClusterConfigService clusterConfigService;
+    private final UserService userService;
+    private final ServerStatus serverStatus;
+    private final EventBus eventBus;
+
+    @Inject
+    public UserSessionTerminationService(MongoDbSessionDAO sessionDao,
+                                         MongoDBSessionService sessionService,
+                                         DefaultSecurityManager securityManager,
+                                         ClusterConfigService clusterConfigService,
+                                         UserService userService,
+                                         ServerStatus serverStatus,
+                                         EventBus eventBus) {
+        this.sessionDao = sessionDao;
+        this.securityManager = securityManager;
+        this.clusterConfigService = clusterConfigService;
+        this.sessionService = sessionService;
+        this.userService = userService;
+        this.serverStatus = serverStatus;
+        this.eventBus = eventBus;
+    }
+
+    // Listens on the event bus to terminate users sessions when a user gets disabled or deleted.
+    // The UserChangedEvent is only published on the local EventBus and not the cluster EventBus so every node
+    // should react to change events.
+    @Subscribe
+    public void handleUserChanged(UserChangedEvent event) {
+        final User user = userService.loadById(event.userId());
+
+        if (user != null && SESSION_TERMINATION_STATUS.contains(user.getAccountStatus())) {
+            terminateSessionsForUser(user);
+        }
+    }
+
+    @Override
+    protected void startUp() throws Exception {
+        try {
+            runGlobalSessionTermination();
+        } catch (Exception e) {
+            LOG.error("Couldn't run global session termination", e);
+        }
+
+        eventBus.register(this);
+    }
+
+    @Override
+    protected void shutDown() throws Exception {
+        eventBus.unregister(this);
+    }
+
+    private boolean isNotPrimaryNode() {
+        return !serverStatus.hasCapability(ServerStatus.Capability.MASTER);
+    }
+
+    // Terminates all active user sessions when the TERMINATION_REVISION has been bumped.
+    private void runGlobalSessionTermination() {
+        if (isNotPrimaryNode()) {
+            LOG.debug("Only run on the primary node to avoid concurrent session termination");
+            return;
+        }
+
+        final GlobalTerminationRevisionConfig globalTerminationRevisionConfig = clusterConfigService.getOrDefault(
+                GlobalTerminationRevisionConfig.class,
+                GlobalTerminationRevisionConfig.initial()
+        );
+
+        if (!globalTerminationRevisionConfig.isOutdated()) {
+            LOG.debug("Global session termination not required");
+            return;
+        }
+
+        final AtomicLong terminatedSessions = new AtomicLong(0);
```

Comment:
Out of curiosity: what's the reason to use an `AtomicLong` instead of a primitive type here?

Reply:
@thll Haha, lol. This is a leftover from a previous version where I incremented the long inside a lambda. I will update it to use a primitive. Good catch, thanks!